File: im-mobile/lib/stores/scheduled_message_store.dart
import 'package:im_mobile/models/scheduled_message_model.dart';
import 'package:im_mobile/services/api_service.dart';
import 'package:mobx/mobx.dart';

part 'scheduled_message_store.g.dart';

class ScheduledMessageStore = _ScheduledMessageStore with _$ScheduledMessageStore;

abstract class _ScheduledMessageStore with Store {
  final ApiService _apiService = ApiService();

  @observable
  ObservableList<ScheduledMessageModel> messages = ObservableList<ScheduledMessageModel>();

  @observable
  bool loading = false;

  @observable
  String? error;

  @observable
  int totalCount = 0;

  @observable
  int pendingCount = 0;

  @observable
  int page = 0;

  @observable
  int pageSize = 20;

  @observable
  ScheduledMessageStatus? filterStatus;

  @action
  Future<void> fetchMessages({bool reset = false}) async {
    if (reset) {
      page = 0;
      messages.clear();
    }

    loading = true;
    error = null;

    try {
      final params = <String, dynamic>{
        'page': page,
        'size': pageSize,
      };

      if (filterStatus != null) {
        params['status'] = filterStatus!.name;
      }

      final response = await _apiService.get('/scheduled-messages', queryParameters: params);
      
      final List<dynamic> data = response.data['data'];
      final newMessages = data.map((json) => ScheduledMessageModel.fromJson(json)).toList();

      if (reset) {
        messages = ObservableList.of(newMessages);
      } else {
        messages.addAll(newMessages);
      }
      
      totalCount = response.data['totalElements'];
      loading = false;
    } catch (e) {
      error = '获取定时消息失败: $e';
      loading = false;
    }
  }

  @action
  Future<bool> createMessage({
    required int receiverId,
    required String content,
    required DateTime scheduledTime,
  }) async {
    loading = true;
    error = null;

    try {
      final response = await _apiService.post('/scheduled-messages', data: {
        'receiverId': receiverId,
        'content': content,
        'scheduledTime': scheduledTime.toIso8601String(),
      });

      final newMessage = ScheduledMessageModel.fromJson(response.data['data']);
      messages.insert(0, newMessage);
      totalCount++;
      loading = false;
      return true;
    } catch (e) {
      error = '创建定时消息失败: $e';
      loading = false;
      return false;
    }
  }

  @action
  Future<bool> cancelMessage(int messageId) async {
    try {
      final response = await _apiService.post('/scheduled-messages/$messageId/cancel');
      final updatedMessage = ScheduledMessageModel.fromJson(response.data['data']);
      
      final index = messages.indexWhere((m) => m.id == messageId);
      if (index != -1) {
        messages[index] = updatedMessage;
      }
      
      return true;
    } catch (e) {
      error = '取消定时消息失败: $e';
      return false;
    }
  }

  @action
  Future<bool> deleteMessage(int messageId) async {
    try {
      await _apiService.delete('/scheduled-messages/$messageId');
      messages.removeWhere((m) => m.id == messageId);
      totalCount--;
      return true;
    } catch (e) {
      error = '删除定时消息失败: $e';
      return false;
    }
  }

  @action
  Future<void> fetchStats() async {
    try {
      final response = await _apiService.get('/scheduled-messages/stats');
      pendingCount = response.data['data']['pendingCount'];
    } catch (e) {
      print('获取统计信息失败: $e');
    }
  }

  @action
  void setFilterStatus(ScheduledMessageStatus? status) {
    filterStatus = status;
    fetchMessages(reset: true);
  }

  @action
  void loadMore() {
    if (messages.length < totalCount && !loading) {
      page++;
      fetchMessages();
    }
  }

  @action
  void clearError() {
    error = null;
  }
}

File: im-desktop/src/stores/scheduled-message-store.ts
import { makeAutoObservable, runInAction } from 'mobx';
import { ScheduledMessage, ScheduledMessageStatus } from '../types/scheduled-message';
import { api } from '../utils/api';

/**
 * 定时消息状态管理
 */
class ScheduledMessageStore {
  messages: ScheduledMessage[] = [];
  loading: boolean = false;
  error: string | null = null;
  currentMessage: ScheduledMessage | null = null;
  totalCount: number = 0;
  pendingCount: number = 0;
  page: number = 0;
  pageSize: number = 20;
  filterStatus: ScheduledMessageStatus | null = null;

  constructor() {
    makeAutoObservable(this);
  }

  /**
   * 获取定时消息列表
   */
  async fetchMessages(reset: boolean = false) {
    if (reset) {
      this.page = 0;
      this.messages = [];
    }

    this.loading = true;
    this.error = null;

    try {
      const params: Record<string, any> = {
        page: this.page,
        size: this.pageSize,
      };

      if (this.filterStatus) {
        params.status = this.filterStatus;
      }

      const response = await api.get('/scheduled-messages', { params });

      runInAction(() => {
        if (reset) {
          this.messages = response.data.data;
        } else {
          this.messages = [...this.messages, ...response.data.data];
        }
        this.totalCount = response.data.totalElements;
        this.loading = false;
      });
    } catch (err: any) {
      runInAction(() => {
        this.error = err.response?.data?.message || '获取定时消息失败';
        this.loading = false;
      });
    }
  }

  /**
   * 加载更多
   */
  loadMore() {
    if (this.messages.length < this.totalCount && !this.loading) {
      this.page += 1;
      this.fetchMessages();
    }
  }

  /**
   * 创建定时消息
   */
  async createMessage(message: Omit<ScheduledMessage, 'id' | 'status' | 'createdAt'>): Promise<boolean> {
    this.loading = true;
    this.error = null;

    try {
      const response = await api.post('/scheduled-messages', message);

      runInAction(() => {
        this.messages.unshift(response.data.data);
        this.totalCount += 1;
        this.loading = false;
      });

      return true;
    } catch (err: any) {
      runInAction(() => {
        this.error = err.response?.data?.message || '创建定时消息失败';
        this.loading = false;
      });
      return false;
    }
  }

  /**
   * 更新定时消息
   */
  async updateMessage(id: number, message: Partial<ScheduledMessage>): Promise<boolean> {
    this.loading = true;
    this.error = null;

    try {
      const response = await api.put(`/scheduled-messages/${id}`, message);

      runInAction(() => {
        const index = this.messages.findIndex(m => m.id === id);
        if (index !== -1) {
          this.messages[index] = response.data.data;
        }
        this.loading = false;
      });

      return true;
    } catch (err: any) {
      runInAction(() => {
        this.error = err.response?.data?.message || '更新定时消息失败';
        this.loading = false;
      });
      return false;
    }
  }

  /**
   * 取消定时消息
   */
  async cancelMessage(id: number): Promise<boolean> {
    try {
      const response = await api.post(`/scheduled-messages/${id}/cancel`);

      runInAction(() => {
        const index = this.messages.findIndex(m => m.id === id);
        if (index !== -1) {
          this.messages[index] = response.data.data;
        }
      });

      return true;
    } catch (err: any) {
      runInAction(() => {
        this.error = err.response?.data?.message || '取消定时消息失败';
      });
      return false;
    }
  }

  /**
   * 删除定时消息
   */
  async deleteMessage(id: number): Promise<boolean> {
    try {
      await api.delete(`/scheduled-messages/${id}`);

      runInAction(() => {
        this.messages = this.messages.filter(m => m.id !== id);
        this.totalCount -= 1;
      });

      return true;
    } catch (err: any) {
      runInAction(() => {
        this.error = err.response?.data?.message || '删除定时消息失败';
      });
      return false;
    }
  }

  /**
   * 获取统计信息
   */
  async fetchStats() {
    try {
      const response = await api.get('/scheduled-messages/stats');
      runInAction(() => {
        this.pendingCount = response.data.data.pendingCount;
      });
    } catch (err) {
      console.error('获取统计信息失败', err);
    }
  }

  /**
   * 设置状态筛选
   */
  setFilterStatus(status: ScheduledMessageStatus | null) {
    this.filterStatus = status;
    this.fetchMessages(true);
  }

  /**
   * 清除错误
   */
  clearError() {
    this.error = null;
  }
}

export const scheduledMessageStore = new ScheduledMessageStore();
